chore: extract `TestEnv` into separate crate

LagginTimes opened the pull request:
### Description

`TestEnv` is extracted into its own crate with `electrsd` support added so that `TestEnv` can also serve `esplora` and `electrum`.
The tests in the `esplora` crate have also been updated to use `TestEnv`.

The `tx_can_become_unconfirmed_after_reorg()` test in `test_electrum` suggests that the electrum issue where a reorged tx would be stuck at confirmed does not exist anymore.

### Notes to the reviewers

The code for `tx_can_become_unconfirmed_after_reorg()` was adapted from the same test in `bitcoind_rpc/test_emitter`. This electrum version of the test requires extra review, as I am uncertain if I used the API efficiently.

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

evanlinjin commented:
The commit message does not need `()`.

evanlinjin reviewed (changes requested):
We need to change the internals to use `electrsd` so the test env can serve `esplora` and `electrum` as well.

Can use https://github.com/bitcoindevkit/bdk/pull/979 as a reference to see what electrum-server methods we need.

LagginTimes commented:
> We need to change the internals to use `electrsd` so the test env can serve `esplora` and `electrum` as well.

I will look into this().

vladimirfomene commented:
I think having reorg functions for all backends will help. I had to use a hack from Steve to implement reorg for electrum. You should see the code in #979. Read more here: https://bitcoin.stackexchange.com/questions/114044/how-can-i-simulate-a-reorg-for-testing.

evanlinjin commented:
@vladimirfomene that's a great point.

https://github.com/bitcoindevkit/bdk/blob/release/0.29/src/testutils/blockchain_tests.rs

Maybe the old `testutils` is something to base this work on.

evanlinjin commented on crates/bitcoind_rpc/tests/test_emitter.rs:
```diff
@@ -189,15 +43,23 @@ fn block_to_chain_update(block: &bitcoin::Block, height: u32) -> local_chain::Up
-        hashes.extend(env.mine_blocks(101, None)?);
+        let mut hashes = (0..=tip)
+            .map(|height| env.rpc_client().get_block_hash(height))
+            .collect::<Result<Vec<_>, _>>()?; // includes genesis block
```
Can we leave a comment explaining why we need to do this? Thanks!

evanlinjin commented on crates/bitcoind_rpc/tests/test_emitter.rs:
```diff
@@ -189,15 +43,23 @@ fn block_to_chain_update(block: &bitcoin::Block, height: u32) -> local_chain::Up
+    (0..tip).for_each(|height| {
+        let changeset = BTreeMap::from([(height as u32, Some(exp_hashes[height as usize]))]);
+        local_chain.apply_changeset(&changeset);
+    });
```
This also needs a comment saying why.

Nit: I would iterate over `exp_hashes` (using `.enumerate` and `.take`). I would also use the `LocalChain::from_blocks` method to construct the `local_chain` here instead.

evanlinjin commented on crates/bitcoind_rpc/Cargo.toml:
```diff
@@ -12,6 +12,7 @@ bitcoincore-rpc = { version = "0.17" }
 
 [dev-dependencies]
+testenv = { path = "../testenv", version = "0.1.0", default_features = false }
 bitcoind = { version = "0.33", features = ["25_0"] }
```
We can get rid of this test dependency now.

evanlinjin commented on crates/testenv/src/lib.rs:
```diff
@@ -0,0 +1,189 @@
+    pub fn invalidate_blocks(
+        &self,
+        count: usize,
+        bitcoind: &electrsd::bitcoind::BitcoinD,
```
Why do we need this as a parameter when we already have a `BitcoinD` in `self`?

Edit: Okay I see your intention here. Concept ACK, Approach NACK. Let's have two `bitcoinD`s in `TestEnv` instead.

LagginTimes replied:
It seems that the simple electrsd reorg test completed without having to pass in `bitcoinD` references. I have removed the extraneous `bitcoind` parameters.

evanlinjin commented on crates/testenv/src/lib.rs:
```diff
@@ -0,0 +1,189 @@
+        res
+    }
+
+    pub fn reorg_empty_blocks(&self, count: usize) -> anyhow::Result<Vec<(usize, BlockHash)>> {
```
Don't we also need to make this method `electrs`-safe?

evanlinjin commented on crates/testenv/src/lib.rs:
Please add documentation, which should include important context as to why some things are implemented in a certain way.

I.e. The requirement of https://github.com/bitcoindevkit/bdk/pull/1171#issuecomment-1760465589 would not be immediately apparent, so a note of this would be mandatory.

evanlinjin replied:
Please add documentation to all public methods and structures.

evanlinjin reviewed (changes requested):
Thanks for taking this forward. The code looks pretty good.

The documentation was a little lacking. It's important to communicate why things are implemented the way they are.

LagginTimes commented:
A simple electrsd reorg test was written and runs without having to pass in `bitcoinD` references to `mine_blocks()` and `reorg()`. I have therefore removed the extraneous `bitcoind` input parameters from `TestEnv`. Will have this commit sorted out tomorrow.

LagginTimes commented:
A `test_chain_sync()` test has been added that reproduces the issue described in #1125 and also confirms that the fix in #1145 at least partially fixes the problem (in that there may be additional bugs as seen by @danielabrozzoni).

evanlinjin commented on crates/bitcoind_rpc/Cargo.toml:
```diff
@@ -19,7 +19,7 @@ bitcoincore-rpc = { version = "0.17" }
 
 [dev-dependencies]
-bitcoind = { version = "0.33", features = ["25_0"] }
+testenv = { path = "../testenv", version = "0.1.0", default_features = false }
```
Let's not include the `version` here.

LLFourn replied:
Why wouldn't you include the version? I think you need it to publish on crates.io

evanlinjin replied:
This is a dev dependency. So when we release a new version of `test_env` we won't have to change all the version fields of these dev dependencies. Less things to change for a release

evanlinjin commented on crates/electrum/Cargo.toml:
```diff
@@ -12,6 +12,11 @@ readme = "README.md"
 #rustls = { version = "=0.21.1", optional = true, features = ["dangerous_configuration"] }
+
+[dev-dependencies]
+testenv = { path = "../testenv", version = "0.1.0", default-features = false }
```
Please remove the `version` field.

evanlinjin commented on crates/electrum/Cargo.toml:
```diff
@@ -12,6 +12,11 @@ readme = "README.md"
 
 [dependencies]
-bdk_chain = { path = "../chain", version = "0.6.0", default-features = false }
+bdk_chain = { path = "../chain", version = "0.6.0", features = [ "miniscript" ] }
```
This will force the crate to use `bdk_chain` with default features and `"miniscript"` when it's not strictly needed.

Can you revert this and add a `bdk_chain` dependency under `[dev-dependencies]`?

LLFourn replied:
Why do we need miniscript to test this crate?

LagginTimes replied:
Including `miniscript` was to get a very early version of `test_electrum` to work properly. It is no longer needed, but I forgot to remove it.

evanlinjin commented on crates/esplora/Cargo.toml:
```diff
@@ -21,6 +21,9 @@ futures = { version = "0.3.26", optional = true }
 miniscript = { version = "10.0.0", optional = true, default-features = false }
 
+[dev-dependencies]
+testenv = { path = "../testenv", version = "0.1.0", default_features = false }
```
rm `version` field thanks!

evanlinjin commented on crates/testenv/Cargo.toml:
```diff
@@ -0,0 +1,19 @@
+
+[dependencies]
+# For no-std, remember to enable the bitcoin/no-std feature
+bitcoin = { version = "0.30", default-features = false }
```
`bdk_chain` already re-exports `bitcoin` so we don't need this.

evanlinjin commented on crates/bitcoind_rpc/tests/test_emitter.rs:
```diff
@@ -188,13 +42,17 @@ fn block_to_chain_update(block: &bitcoin::Block, height: u32) -> local_chain::Up
+    let (mut local_chain, _) = LocalChain::from_genesis_hash(env.rpc_client().get_block_hash(0)?);
+    let mut emitter = Emitter::new(env.rpc_client(), local_chain.tip(), 0);
 
     // mine some blocks and returned the actual block hashes
```
nit

```suggestion
    // mine some blocks and returned the actual block hashes
    // because creating an instance of `ElectrsD` automatically mines blocks... we need to... 
```

evanlinjin commented on crates/electrum/tests/test_electrum.rs:
```diff
@@ -0,0 +1,238 @@
+        let graph_update =
+            relevant_txids.into_confirmation_time_tx_graph(&client, None, missing)?;
+        let _ = recv_chain.apply_update(chain_update);
+        let _ = recv_graph.apply_update(graph_update);
```
Can we check the changeset here to see if a new anchor is added?

ValuedMammal replied:
> Can we check the changeset here to see if a new anchor is added?

We would expect to see new anchors inserted into graph when the initially sent txns confirm, but what would be the expectation for anchors *after* reorg?

evanlinjin commented on crates/testenv/src/lib.rs:
```diff
@@ -0,0 +1,198 @@
+                    &electrsd::bitcoind::Conf::default(),
+                )
+            }
+        }?;
```
We only need one ENV here. For consistency, use `BITCOIND_EXE` and scrap `TEST_BITCOIND`.

evanlinjin commented on crates/testenv/src/lib.rs:
```diff
@@ -0,0 +1,198 @@
+        Ok(block_hashes)
+    }
+
+    /// Mine an empty block.
```
```suggestion
    /// Mine a block that is guaranteed to be empty even with transactions in the mempool.
```

evanlinjin reviewed (changes requested):
Very close!

evanlinjin commented:
In terms of commit-hygiene, I think this PR should be at least two commits. One for extracting the `TestEnv`, one for introducing the `electrum` tests.

ValuedMammal reviewed (commented):
Light review ACK, hoping to do a deeper dive on `test_electrum.rs` soon.

update: tested Ack. I don't see any major issues. Great work @LagginTimes

update: made a comment on 53186e2a4d5c4a2e71580ea27679de99f5cc4782

ValuedMammal commented on crates/electrum/tests/test_electrum.rs:
```diff
@@ -0,0 +1,240 @@
+
+    let missing = relevant_txids.missing_full_txs(recv_graph.graph());
+    let graph_update = relevant_txids.into_confirmation_time_tx_graph(&client, None, missing)?;
+    let _ = recv_chain.apply_update(chain_update);
```
```suggestion
    let chain_changeset = recv_chain.apply_update(chain_update).unwrap();
    assert_eq!(chain_changeset.len(), REORG_COUNT);
```

This should be true, but right now `chain_changeset` only contains 8 elements. I think it's because we only fetch `CHAIN_SUFFIX_LENGTH` number of blocks (8) from electrum, which adds an extra wrinkle I'm not sure was intended.

If we set `REORG_COUNT` to 8, then this is resolved, and we won't find new anchors added post-reorg.

LagginTimes replied:
Good catch. Fixed.

ValuedMammal commented on crates/electrum/tests/test_electrum.rs:
```diff
@@ -0,0 +1,240 @@
+
+    let missing = relevant_txids.missing_full_txs(recv_graph.graph());
+    let graph_update = relevant_txids.into_confirmation_time_tx_graph(&client, None, missing)?;
+    let _ = recv_chain.apply_update(chain_update);
```
I'm tempted to use `?` here, to check the Result of `LocalChain::apply_update` is ok. Not sure if you have an opinion, or if this matters. The same suggestion would apply to lines 190, 219 below.

evanlinjin commented on crates/bitcoind_rpc/tests/test_emitter.rs:
```diff
@@ -166,17 +20,22 @@ impl TestEnv {
     let env = TestEnv::new()?;
-    let (mut local_chain, _) = LocalChain::from_genesis_hash(env.client.get_block_hash(0)?);
-    let mut emitter = Emitter::new(&env.client, local_chain.tip(), 0);
+    let tip = env.rpc_client().get_block_count()?;
```
Nit: 

```suggestion
    let network_tip = env.rpc_client().get_block_count()?;
```

evanlinjin commented on crates/bitcoind_rpc/tests/test_emitter.rs:
```diff
@@ -166,17 +20,22 @@ impl TestEnv {
-    // mine some blocks and returned the actual block hashes
+    // Mine some blocks and return the actual block hashes.
+    // We must first add the hashes of the genesis block and the block automatically generated from
+    // initializing `ElectrsD`.
```
Nit: I think this is clearer?

```suggestion
    // Mine some blocks and return the actual block hashes.
    // Because initializing `ElectrsD` already mines some blocks, we must include those too when returning block hashes.
```

evanlinjin commented on crates/electrum/tests/test_electrum.rs:
```diff
@@ -0,0 +1,240 @@
+fn scan_detects_confirmed_tx() -> Result<()> {
+    const SEND_AMOUNT: Amount = Amount::from_sat(10_000);
+
+    let env = TestEnv::new()?;
+    let client = electrum_client::Client::new(env.electrsd.electrum_url.as_str())?;
```
Now #1306 is merged, I think we can just use `env.electrum_client()`.

evanlinjin replied:
Getting the versions of the crates to match up right now is impossible. I think what you have is optimal for now.

evanlinjin commented on crates/electrum/tests/test_electrum.rs:
```diff
@@ -0,0 +1,240 @@
+        }
+        std::thread::sleep(delay);
+    }
+}
```
I think this can be a method on `TestEnv`. Maybe rename it to `wait_until_electrum_sees_block` or something more descriptive. It's also not immediately apparent why this is useful so please write some docs!

evanlinjin commented on crates/testenv/Cargo.toml:
```diff
@@ -0,0 +1,19 @@
+[package]
+name = "testenv"
```
Please rename to `bdk_testenv`.

evanlinjin commented on crates/testenv/src/lib.rs:
```diff
@@ -0,0 +1,196 @@
+        res
+    }
+
+    /// Reorg with a number of empty blocks of a given size `count`.
```
Docs can be a bit more descriptive here. I.e. will mine empty blocks even though there are txs in the mempool. Or: refer to `mine_empty_blocks` for more.

evanlinjin reviewed (changes requested):
This is looking great. This is my initial round of reviews. I still need to look at the introduced electrum tests.

evanlinjin commented:
I think you will need to do a really annoying rebase :/ sorry about that!

notmandatory commented:
This one needs a rebase after alpha.5 release.

LagginTimes committed:
chore: extract `TestEnv` into separate crate
`TestEnv` is extracted into its own crate to serve as a framework
for testing other block explorer APIs.

LagginTimes committed:
test(electrum): added scan and reorg tests
Added scan and reorg tests to check electrum functionality using
`TestEnv`.

evanlinjin commented:
I think git is being weird, can you re-rebase on master? Make sure your remote is up to date!

evanlinjin commented:
Also could you please add a simple README for `bdk_testenv`? Thanks!

ValuedMammal commented on crates/testenv/src/lib.rs:
```diff
@@ -0,0 +1,233 @@
+                electrsd_conf.http_enabled = true;
+                let electrs_exe = electrsd::downloaded_exe_path()
+                    .expect("electrs version feature must be enabled");
+                electrsd::ElectrsD::with_conf(electrs_exe, &bitcoind, &electrsd_conf)
```
If `ELECTRS_EXE` is Some, we're calling `new` with no `electrsd::Conf`. We should configure `with_conf` in both cases.

oleonardolima replied:
I think that in the background it's already doing that, right ? 🤔 https://docs.rs/electrsd/0.25.0/src/electrsd/lib.rs.html#143-146

But I agree that a standard form would be better :)

ValuedMammal commented on crates/testenv/src/lib.rs:
```diff
@@ -0,0 +1,233 @@
+                electrsd_conf.http_enabled = true;
+                let electrs_exe = electrsd::downloaded_exe_path()
+                    .expect("electrs version feature must be enabled");
+                electrsd::ElectrsD::with_conf(electrs_exe, &self.bitcoind, &electrsd_conf)
```
If `ELECTRS_EXE` is Some, we're calling `new` with no `electrsd::Conf`. We should configure `with_conf` in both cases.

storopoli commented:
I know that we're testing it in other crates.
But do we plan to expose this in the future?
If yes, I think we should add tests for the `test_env` crate (I know that this sounds like an oxymoron, but bear with me ...)

ValuedMammal commented:
Now that #1351 is open, I made LagginTimes/bdk#1 to test stop_gap in `ElectrumExt::full_scan`

danielabrozzoni commented on Cargo.toml:
```diff
@@ -7,7 +7,7 @@ members = [
     "crates/esplora",
     "crates/bitcoind_rpc",
-    "crates/hwi",
+    "crates/testenv",
```
Why are we removing hwi from the Cargo.toml?

evanlinjin replied:
Damn I didn't catch this

evanlinjin replied:
@LagginTimes is it because something is failing to build?

LagginTimes replied:
I believe this may have been an accidental deletion as a result of a merge. Thanks for catching this.

evanlinjin reviewed (approved):
ACK 29036977e11696007825b49c894c8b06d92337b2

notmandatory commented on .github/workflows/nightly_docs.yml:
```diff
@@ -17,6 +17,8 @@ jobs:
         run: rustup update
       - name: Rust Cache
         uses: Swatinem/rust-cache@v2.2.1
+      - name: Pin dependencies for MSRV
```
This shouldn't be needed. I was able to build the locally docs (`cargo doc --no-deps`) with `nightly-2022-12-14` without any errors.

LagginTimes replied:
Publish Nightly Docs / Build docs will unfortunately fail without this change.

> error: package `home v0.5.9` cannot be built because it requires rustc 1.70.0 or newer, while the currently active rustc version is 1.68.0-nightly
Either upgrade to rustc 1.70.0 or newer, or use
cargo update -p home@0.5.9 --precise ver
where `ver` is the latest version of `home` supporting rustc 1.68.0-nightly
Error: Process completed with exit code 101.

notmandatory replied:
That's strange, but anyway not a big deal, more important to get the rest of this PR in. Thanks for checking.

notmandatory commented on crates/testenv/Cargo.toml:
```diff
@@ -0,0 +1,19 @@
+
+[dependencies]
+# For no-std, remember to enable the bitcoin/no-std feature
+# bitcoin = { version = "0.30", default-features = false }
```
Does this commented line need to be here? I don't see how it is related to the above comment.

notmandatory reviewed (commented):
Overall looks like a great improvement, related tests are much easier to read too. I just have one comment/question about a CI change.

LagginTimes committed:
chore: update MSRV dependency for nightly docs

storopoli reviewed (approved):
ACK e8dc8d3d7dd30aa3b0826b6ad50399a57a105491

oleonardolima commented on crates/electrum/tests/test_electrum.rs:
```diff
@@ -0,0 +1,225 @@
+///
+/// 1. Mine 101 blocks.
+/// 2. Mine 11 blocks with a confirmed tx in each.
+/// 3. Perform 11 separate reorgs on each block with a confirmed tx.
```
@LagginTimes Is this documentation up-to-date, shouldn't it be 8 as it's on `REORG_COUNT` ?

oleonardolima replied:
I mean, [here](https://github.com/bitcoindevkit/bdk/pull/1171/commits/ac4345462ef01c0d80210b6e4038483ce83b4685#diff-6febce892f6a2b6f549cd1fec8fcff9f5512bf62417685b073c53a6a0b252c6cR156-R160) you are creating only 8txs that will take part in the reorganization later, right?

LagginTimes replied:
> I mean, [here](https://github.com/bitcoindevkit/bdk/pull/1171/commits/ac4345462ef01c0d80210b6e4038483ce83b4685#diff-6febce892f6a2b6f549cd1fec8fcff9f5512bf62417685b073c53a6a0b252c6cR156-R160) you are creating only 8txs that will take part in the reorganization later, right?

This is correct, the docs weren't updated at the time of the change to 8.

oleonardolima commented on crates/bitcoind_rpc/tests/test_emitter.rs:
```diff
@@ -744,7 +624,7 @@ fn mempool_during_reorg() -> anyhow::Result<()> {
         // emission.
         // TODO: How can have have reorg logic in `TestEnv` NOT blacklast old blocks first?
```
@LagginTimes general question: after this PR is merged, do you think we'll be able to solve this TODO ?

evanlinjin replied:
Unfortunately no. This can be done in a follow-up PR!

oleonardolima replied:
Cool, thanks!

oleonardolima reviewed (commented):
Overall looks pretty good!

I just left one minor comment about the documentation that I was uncertain about, and a general question for further discussion.

evanlinjin commented on crates/esplora/Cargo.toml:
```diff
@@ -21,6 +21,9 @@ futures = { version = "0.3.26", optional = true }
+bdk_testenv = { path = "../testenv", default_features = false }
+
 [target.'cfg(not(target_arch = "wasm32"))'.dev-dependencies]
 electrsd = { version= "0.25.0", features = ["bitcoind_25_0", "esplora_a33e97e1", "legacy"] }
```
Do we need to make the tests run in WASM? Because `electrsd` is a mandatory dependency of `bdk_testenv`.

notmandatory replied:
Right now for CI we only do a `cargo check` for WASM and don't rerun the tests for that target. Since I haven't heard of any WASM issues that slipped through by not running the tests I don't think we need to change `TestEnv` to work with WASM.

oleonardolima reviewed (approved):
ACK 04d0ab5a978fbfd45a5c795f9142ff4a9ba29ee3

notmandatory reviewed (approved):
ACK 04d0ab5a978fbfd45a5c795f9142ff4a9ba29ee3

Great job, this will be a big help with testing.

evanlinjin committed:
chore(esplora): rm custom WASM target spec test dependency

None of the `bdk_esplora` tests can run under WASM anyway since we
depend on `electrsd`.

evanlinjin committed:
chore(electrum,testenv): move `test_reorg_is_detected_in_electrsd`

evanlinjin reviewed (approved):
ACK 7c9ba3cfc855dc0845476fe923dac4e47cb5c06a
